# Домашнє завдання до теми "Python Basics"
В цьому ДЗ важливою умовою виконання завдань є те, що ми не імпортуємо жодних бібліотек, а робимо все на чистому Python, використовуючи лише базовий функціонал. Це необхідно аби відпрацювати роботу з базовими типами даних. А використовувати бібліотеки ми ще навчимось далі :)

# Робота з рядками

1.Запишіть у змінну `name` ваше імʼя.

In [1]:
name = 'Dmytro'

2.Запишіть у змінну `surname` ваше прізвище.

In [2]:
surname = 'Horokhov'

3.Сконкатенуйте імʼя та прізвище в одну фразу через пробіл. Запишіть результат у нову змінну `name_surname` та виведіть результат:

In [8]:
name_surname = name + ' ' + surname
name_surname

'Dmytro Horokhov'

Очікувана відповідь (тільки з вашими даними): 'Hanna Pylieva' 

4.Зробіть усі літери змінної `name_surname` заглавними використовуючи вбудовану функцію Python для рядків:

In [6]:
name_surname.upper()

'DMYTRO HOROKHOV'

Очікувана відповідь (тільки з вашими даними): 'HANNA PYLIEVA'

5.Виведіть третю літеру вашого `name_surname`. Тут уважно, треба саме третя літера. Згадайте, з якого значення починається індексування значень в Python.

In [12]:
name_surname[2]

'y'

Очікувана відповідь (тільки з вашими даними): 'n'

6.Виведіть першу та останню літеру вашого прізвища.

In [15]:
name_surname[0], name_surname[-1]

('D', 'v')

Очікувана відповідь (тільки з вашими даними): ('P', 'a')

7.Виведіть довжину вашого прізвища.

In [16]:
len(surname)

8

Очікувана відповідь (тільки з вашими даними): 7

8.Виведіть кількість літер `a` у столиці Мадагаскару `Antananarivo`. Важливо врахувати всі літери, як великі, так і маленькі.

In [24]:
Capital = 'Antananarivo'
print(Capital.count('A') + Capital.count('a'))

4


Очікувана відповідь: 4

# Робота з числами та операторами

9.Створіть змінні a, b, c та d цілочисельного типу, задайде їм значення відповідно 1, 2, 3, 4, та обчисліть з допомогою Python наступний вираз:  $$ \frac{(a+b) * c}{d} $$

In [25]:
a = 1
b = 2
c = 3
d = 4
(a + b) * c / d

2.25

Очікувана відповідь: 2.25

10.На змінних, створених в попердньому завданні, обчисліть наступний вираз з допомогою Python: $$ (a^b + c^d)^2 $$

In [26]:
(a**b + c**d)**2

6724

Очікувана відповідь: 6724

# Робота зі колекціями

У нас є наступний список:

In [108]:
l = [
    78, 12, 14, 27, 47, 28, 51, 46, 94, 37, 42, 20, 52, 33, 11, 11, 51,
    20, 81, 54, 86, 51, 38, 42, 49, 99, 76, 17,  9, 23, 97, 29, 29, 63,
    40, 11, 31, 76, 15, 21, 29, 75, 35,  2, 10, 71, 17, 10, 13,  7, 44,
    12, 19, 61, 13, 25, 62, 46, 10, 44, 75, 57, 20, 57,  3, 44, 16, 28,
    79, 57, 62, 70, 95,  8, 47, 82, 80, 45, 36, 44, 79, 95, 73, 42, 29,
    83, 80, 66, 56,  8, 73,  5, 99, 65, 65, 98, 17, 36, 16, 60
]

Будемо з ним працювати в наступних завданнях.

11.Виведіть 5 перших елементів списку

In [28]:
l[:5]

[78, 12, 14, 27, 47]

Очікувана відповідь: [78, 12, 14, 27, 47]

12.Виведіть максимальний і мінімальний елемент у списку

In [42]:
print('Max element', max(l), '\n', 'Min element', min(l))

Max element 99 
 Min element 2


Очікувана відповідь: 

```
Max element  99
Max element  2
```

13.Чи є в списку елементи, які повторюються? Перевірте це програмно і з допомогою інструкцій `if ... else` виведіть на екран відповідь "Так, є елементи, які повторюються", або "Ні, всі елементи унікальні".
Важливо це зробити без використання імпортів будь-яких бібліотек, на чистому Python так би мовити.

In [53]:
for i in range(len(l)):
    if l.count(l[i]) > 1:
        print("Так, є елементи, які повторюються")
        break
    else:
        continue
    print("Ні, всі елементи унікальні")

Так, є елементи, які повторюються


14.Для кожного унікального елементу списку `l` знайдіть, скільки разів він повторюється та запишіть результат у змінну `n_repeats`. Виведіть `n_repeats` на екран.

In [94]:
n_repeats = {}
for x in set(l):
    n_repeats[x] = l.count(x)
print(n_repeats)

{2: 1, 3: 1, 5: 1, 7: 1, 8: 2, 9: 1, 10: 3, 11: 3, 12: 2, 13: 2, 14: 1, 15: 1, 16: 2, 17: 3, 19: 1, 20: 3, 21: 1, 23: 1, 25: 1, 27: 1, 28: 2, 29: 4, 31: 1, 33: 1, 35: 1, 36: 2, 37: 1, 38: 1, 40: 1, 42: 3, 44: 4, 45: 1, 46: 2, 47: 2, 49: 1, 51: 3, 52: 1, 54: 1, 56: 1, 57: 3, 60: 1, 61: 1, 62: 2, 63: 1, 65: 2, 66: 1, 70: 1, 71: 1, 73: 2, 75: 2, 76: 2, 78: 1, 79: 2, 80: 2, 81: 1, 82: 1, 83: 1, 86: 1, 94: 1, 95: 2, 97: 1, 98: 1, 99: 2}


Очікуваний результат (три крапки - для скорочення, у вас буде виведений повний набір унікальних елементів): 
```
{2: 1,
 3: 1,
 5: 1,
    ...
 97: 1,
 98: 1,
 99: 2}
```

15.Знайдіть три найменші унікальні елементи в списку `l`.

In [86]:
new_list = []
new_set = set(l)
for i in range(3):
    minimum = min(new_set)
    new_list.append(minimum)
    new_set.remove(minimum)
print(new_list)

[2, 3, 5]


Очікувана відповідь: [2, 3, 5]


16.Використовуючи `n_repeats` скажіть, скільки разів повторюється кожен з трьох найменших елементів в списку.

In [99]:
d = {x: n_repeats[x] for x in new_list}
d

{2: 1, 3: 1, 5: 1}

Очікувана відповідь:  {2: 1, 3: 1, 5: 1}

17.Створіть новий список `l2`, який містить кожен 11-й елемент списку `l`. Виведіть `l2` на екран.  
Памʼятаємо, щоб наприклад, перший 11-и1 елемент списку має індекс 10.

In [116]:
l2 = [l[i] for i in range(10, len(l), 11)]
l2

[42, 51, 29, 2, 13, 44, 80, 66, 16]

Очікуваний результат: `[42, 51, 29, 2, 13, 44, 80, 66, 16]`

18.Створіть новий список `l2_squared`, де кожен елемент списку `l2` є піднесеним до степіня 2. 

In [117]:
l2_squared = [x**2 for x in l2]
l2_squared

[1764, 2601, 841, 4, 169, 1936, 6400, 4356, 256]

Очікуваний результат: `[1764, 2601, 841, 4, 169, 1936, 6400, 4356, 256]`

19.Знайдіть суму всіх елементів списку `l2` та виведіть її на екран.

In [119]:
sum(l2)

343

20.Знайдіть середнє значення елементів списку `l2`, збережіть результат у змінну `l2_mean` та виведіть новостворену змінну та ії тип на екран.

In [120]:
l2_mean = sum(l2) / len(l2)
l2_mean

38.111111111111114

21.Створіть новий список `l2_gt_mean`, за наступним правилом: "якщо цей елемент спиcку `l2` є більше за `l2_mean`, то запишемо 1, інакше 0. Виведіть новий список на екран.

In [126]:
l2_gt_mean = [1 if x > l2_mean else 0 for x in l2]
l2_gt_mean

[1, 1, 0, 0, 0, 1, 1, 1, 0]

Очікуваний результат: `[1, 1, 0, 0, 0, 1, 1, 1, 0]`

# Ураа!! Ви завершили перше домашнє завдання!!!  
Аби здати ДЗ - залийте ноутбук у гітхаб репозиторій, який ви створили у перші дні навчання на курсі, та додайте посилання на файл у відповідному завданні в Eduflow.
Так тримати! 🚀🚀🚀